### PR TITLE
Implemented tail call instructions:

### PR DESCRIPTION
**Pull request by fgmccabe**
return_call
	return_call_indirect

with some simple tests thereof.

### CODE REVIEW

File: test/interp/return-call-indirect.txt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    i32.const 0
+    i32.gt_s
+    if (result i32)
+      get_local 0
+      i32.const 1
+      i32.sub
+      get_local 1
+      get_local 0
+      i32.mul
+      get_local 2
```
Comment (binji) [inline test/interp/return-call-indirect.txt]: local 2 seems to just be forwarding through 0

File: test/interp/return-call-indirect.txt [hunk 2]
```diff
@@ -0,0 +1,58 @@
+;;; TOOL: run-interp
+(module
+  (type $iii_i (func (param i32 i32 i32)(result i32)))
+  (table anyfunc (elem $facInd))
+
+  (func (export "call") (result i32)
```
Comment (binji) [inline test/interp/return-call-indirect.txt]: `call` and `helper` seem unrelated to this test?

File: test/interp/return-call.txt [hunk 3]
```diff
@@ -0,0 +1,50 @@
+;;; TOOL: run-interp
+(module
+  (func (export "call") (result i32)
```
Comment (binji) [inline test/interp/return-call.txt]: same here, are `call` and `helper` needed?

File: src/binary-reader.cc [hunk 4]
```diff
@@ -755,6 +755,29 @@ Result BinaryReader::ReadFunctionBody(Offset end_offset) {
         break;
       }
 
+      case Opcode::ReturnCall: {
+        Index func_index;
+        CHECK_RESULT(ReadIndex(&func_index, "call function index"));
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply src/binary-reader.cc]: Done

File: src/interp/binary-reader-interp.cc [hunk 5]
```diff
@@ -1341,6 +1343,52 @@ wabt::Result BinaryReaderInterp::OnCallIndirectExpr(Index sig_index) {
...
+    CHECK_RESULT(EmitOpcode(Opcode::ReturnCall));
+    CHECK_RESULT(EmitFuncOffset(cast<DefinedFunc>(func), func_index));
+  }
+
+  return wabt::Result::Ok;
+}
+
+wabt::Result BinaryReaderInterp::OnReturnCallIndirectExpr(Index sig_index) {
+  if (module_->table_index == kInvalidIndex) {
+    PrintError("found call_indirect operator, but no table");
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply src/interp/binary-reader-interp.cc]: done

File: src/binary-reader.cc [hunk 6]
```diff
@@ -755,6 +755,29 @@ Result BinaryReader::ReadFunctionBody(Offset end_offset) {
         break;
       }
 
+      case Opcode::ReturnCall: {
+        Index func_index;
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply src/binary-reader.cc]: done

File: test/interp/return-call-indirect.txt (hunk 1 above)
Comment (fgmccabe) [reply test/interp/return-call-indirect.txt]: I don't understand this.
The third argument is the function index, not the same as either the first or third argument.
Maybe I am missing something here?

File: test/interp/return-call-indirect.txt (hunk 2 above)
Comment (fgmccabe) [reply test/interp/return-call-indirect.txt]: removed

File: test/interp/return-call-import.txt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+;;; TOOL: run-interp
+;;; ARGS: --host-print
+(module
+  (import "host" "print" (func $imported (param i32) (result i32)))
+
+  (func (export "f") (result i32)
+    i32.const 42
+    return_call $imported
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply test/interp/return-call-import.txt]: Added an unreachable instruction. Otherwise, not sure how to do this.

File: src/binary-reader.h [hunk 8]
```diff
@@ -236,6 +236,8 @@ class BinaryReaderDelegate {
   virtual Result OnNopExpr() = 0;
   virtual Result OnRethrowExpr() = 0;
   virtual Result OnReturnExpr() = 0;
+  virtual Result OnReturnCallExpr(Index func_index) = 0;
+  virtual Result OnReturnCallIndirectExpr(Index func_index) = 0;
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply src/binary-reader.h]: done

File: test/interp/return-call-indirect.txt (hunk 1 above)
Comment (fgmccabe) [reply test/interp/return-call-indirect.txt]: Ah. You could. But the way I did it was kind of deliberate. I was imagining that I was implementing a functional programming language :)
